ctgan/synthesizers/base.py

```diff
@@ -57,13 +57,38 @@ def wrapper(self, *args, **kwargs):
 
 
 class BaseSynthesizer:
-    """Base class for all default synthesizers of ``CTGAN``.
-
-    This should contain the save/load methods.
-    """
+    """Base class for all default synthesizers of ``CTGAN``."""
 
     random_states = None
 
+    def __getstate__(self):
+        """Remove the ``torch`` generator when pickling."""
+        state = self.__dict__.copy()
+        if (
+            isinstance(self.random_states, tuple) and
+            isinstance(self.random_states[0], np.random.RandomState) and
+            isinstance(self.random_states[1], torch.Generator)
+        ):
+            state['_numpy_random_state'] = self.random_states[0].get_state()
+            state['_torch_random_state'] = self.random_states[1].get_state()
+            state.pop('random_states')
+
+        return state
+
+    def __setstate__(self, state):
+        """Set the ``torch`` generator when loading a model."""
+        if '_numpy_random_state' in state and '_torch_random_state' in state:
```

Comment:
Why didn't this fail in the previous version of SDV?

Reply:
Previous `SDV` version was randomly sampling by default, which means no seed was unless you run `randomize_samples=False`. Here is a snippet to reproduce the error in the previous `SDV` version:
```python
from sdv.demo import load_tabular_demo
from sdv.tabular import CTGAN

data = load_tabular_demo('student_placements_pii')
ctg = CTGAN()

ctg.fit(data)
ctg.sample(10, randomize_samples=False)
ctg.save('test-model.pkl')
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[16], line 1
----> 1 ctg.save('test-model.pkl')

File ~/Projects/SDV/SDV.master/sdv/tabular/base.py:876, in BaseTabularModel.save(self, path)
    873 self._package_versions = get_package_versions(getattr(self, '_model', None))
    875 with open(path, 'wb') as output:
--> 876     cloudpickle.dump(self, output)

File ~/.virtualenvs/SDV.master/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py:55, in dump(obj, file, protocol, buffer_callback)
     45 def dump(obj, file, protocol=None, buffer_callback=None):
     46     """Serialize obj as bytes streamed into file
     47
     48     protocol defaults to cloudpickle.DEFAULT_PROTOCOL which is an alias to
   (...)
     53     compatibility with older versions of Python.
     54     """
---> 55     CloudPickler(
     56         file, protocol=protocol, buffer_callback=buffer_callback
     57     ).dump(obj)

File ~/.virtualenvs/SDV.master/lib/python3.8/site-packages/cloudpickle/cloudpickle_fast.py:632, in CloudPickler.dump(self, obj)
    630 def dump(self, obj):
    631     try:
--> 632         return Pickler.dump(self, obj)
    633     except RuntimeError as e:
    634         if "recursion" in e.args[0]:

TypeError: cannot pickle 'torch._C.Generator' object

```

Reply:
Ah makes sense